### PR TITLE
[FLINK-5969] Add savepoint backwards compatibility tests from 1.2 to 1.3

### PR DESCRIPTION
**Pull request by aljoscha**
The binary savepoints and snapshots in the tests were created on the commit of the Flink 1.2.0 release, so we test backwards compatibility within the Flink 1.2.x line. Once this is approved I'll open another PR that transplants these commits on the master branch (with the binary snapshots/savepoints done on Flink 1.2.0) so that we test migration compatibility between 1.2.0 and what is going to be Flink 1.3.x.

I changed the naming of some existing tests so we now have `*From11MigrationTest` and `*From12MigrationTest` (and one ITCase). Immediately after releasing Flink 1.3.0 we should do the same, i.e. introduce `*From13MigrationTest` and ITCase based on the existing tests.

The unit tests are somewhat straightforward: we feed some data into an operator using an operator test harness, then we do a snapshot. (This is the part that has to be done on the "old" version to generate the binary snapshot that goes into the repo). The actual tests restore an operator form that snapshot and verify the output.

The ITCase is a bit more involved. We have a complete Job of user-functions and custom operators that tries to cover as many state/timer combinations as possible. We start the job and, using accumulators, observe the number of received elements in the sink. Once we get all elements we perform a savepoint and cancel the job. Thus we have all state caused by the elements reflected in our savepoint. This has to be done on the "old" version and the savepoint goes into the repo. The restoring job is instrumented with code that verifies restored state and updates accumulators. We listen on the accumulator changes and cancel the job once we have seen all required verifications.

### CODE REVIEW

**Comment by zentol** [comment]
One thing i was wondering about regarding the added IT cases is whether we can create a single class with the logic for both 1.1 and 1.2. Given that the new IT cases appear to be mostly a copy of the existing tests, shouldn't the only difference be which savepoint we use?

**Comment by aljoscha** [comment]
Thanks @zentol! I addressed your (easily resolvable) comments.

About unifying the functionality of the tests/ITCases, I was planning to to that for the upcoming `*From13` tests but for the `*From11` and `*From12` it's difficult because quite some stuff changed between 1.1 and 1.2 and I want the code to mostly work on all branches. For example, this is the new `*From12` tests rebased on the master branch: https://github.com/aljoscha/flink/tree/jira-5969-backwards-compat-12-13. By transplanting but keeping the snapshots that were generated on Flink 1.2.0 we get backwards compatibility tests while the tests in this PR test compatibility within Flink 1.2.x. 

I'm hoping that for the future we can move to a base class and version specific tests. I'll also think about how much is possible right now and try and simplify/unify.

File: flink-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFrom12MigrationTest.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+		 * The actual paths in this depend on the binary checkpoint so it you update this the paths
+		 * here have to be updated as well.
+		 */
+		@Override
+		public void initializeState(FunctionInitializationContext context) throws Exception {
+			OperatorStateStore stateStore = context.getOperatorStateStore();
+
+			ListState<State<T>> restoredBucketStates = stateStore.getSerializableListState("bucket-states");
+
+			int subtaskIndex = getRuntimeContext().getIndexOfThisSubtask();
```
Comment (zentol) [inline flink-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFrom12MigrationTest.java]: unused variable

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/util/AbstractStreamOperatorTestHarness.java [hunk 2]
```diff
@@ -485,7 +501,6 @@ public OperatorStateHandles snapshot(long checkpointId, long timestamp) throws E
 	 * the operator implements this interface.
 	 */
 	@Deprecated
-	@SuppressWarnings("deprecation")
```
Comment (aljoscha) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/util/AbstractStreamOperatorTestHarness.java]: Seems to be a mistake. Fixing ...

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/StatefulJobSavepointFrom12MigrationITCase.java [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+		private final ValueStateDescriptor<Long> stateDescriptor =
+				new ValueStateDescriptor<Long>("state-name", LongSerializer.INSTANCE);
+
+		@Override
+		public void flatMap(Tuple2<Long, Long> value, Collector<Tuple2<Long, Long>> out) throws Exception {
+			out.collect(value);
+
+			getRuntimeContext().getState(stateDescriptor).update(value.f1);
+
+			assertEquals(value.f1, getRuntimeContext().getState(stateDescriptor).value());
```
Comment (aljoscha) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/StatefulJobSavepointFrom12MigrationITCase.java]: There was a point in debugging this whole thing were I wasn't sure wether state was set correctly, so I introduced this sanity check. Could remove it.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/SavepointMigrationTestBase.java [hunk 4]
```diff
@@ -208,11 +190,37 @@ protected final void restoreAndExecute(
...
 		while (DEADLINE.hasTimeLeft()) {
+
+			// try and get a job result, this will fail if the job already failed. Use this
+			// to get out of this loop
+			JobID jobId = jobSubmissionResult.getJobID();
+			FiniteDuration timeout = FiniteDuration.apply(5, TimeUnit.SECONDS);
+
+			try {
+
+				Future<Object> future = Patterns.ask(clusterClient.getJobManagerGateway().actor(),
```
Comment (aljoscha) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/SavepointMigrationTestBase.java]: Changing

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/StatefulJobSavepointFrom12MigrationITCase.java [hunk 5]
```diff
@@ -0,0 +1,771 @@
...
+					timer.getNamespace(),
+					LongSerializer.INSTANCE,
+					stateDescriptor);
+
+			assertEquals(state.value(), timer.getNamespace());
+			getRuntimeContext().getAccumulator(SUCCESSFUL_PROCESSING_TIME_CHECK_ACCUMULATOR).add(1);
+		}
+	}
+
+
```
Comment (aljoscha) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/StatefulJobSavepointFrom12MigrationITCase.java]: Fixing

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/StatefulJobSavepointFrom12MigrationITCase.java (hunk 3 above)
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/StatefulJobSavepointFrom12MigrationITCase.java]: Feel free to keep it, it doesn't hurt after all. Just wanted to make sure i wasn't missing anything.

File: flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+		TimestampedFileInputSplit split2 =
+				new TimestampedFileInputSplit(10, 2, new Path("test/test2"), 101, 200, null);
+
+		TimestampedFileInputSplit split3 =
+				new TimestampedFileInputSplit(10, 1, new Path("test/test2"), 0, 100, null);
+
+		TimestampedFileInputSplit split4 =
+				new TimestampedFileInputSplit(11, 0, new Path("test/test3"), 0, 100, null);
+
+		final OneShotLatch latch = new OneShotLatch();
```
Comment (aljoscha) [inline flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java]: Yes, it's intended, to ensure that the Reader never actually does any reading. I copied this straight from the old tests but I agree that it's not very apparent. I'll add a comment

File: flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+	/**
+	 * Manually run this to write binary snapshot data. Remove @Ignore to run.
+	 */
+	@Ignore
+	@Test
+	public void writeMonitoringSourceSnapshot() throws Exception {
+
+		File testFolder = tempFolder.newFolder();
+
+		File path = null;
```
Comment (aljoscha) [inline flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java]: Also leftovers, removing...

File: flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+		public void close() {
+		}
+	}
+
+	/**
+	 * Create a file with pre-determined String format of the form:
+	 * {@code fileIdx +": "+ sampleLine +" "+ lineNo}.
+	 * */
+	private Tuple2<File, String> createFileAndFillWithData(
+		String base, String fileName, int fileIdx, String sampleLine) throws IOException {
```
Comment (aljoscha) [inline flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java]: Fixing, what's up with the `[.]`?

File: flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java (hunk 6 above)
Comment (zentol) [inline flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java]: But when does the Reader get unblocked? Or do we just destroy it somewhere?

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+				catch (Throwable t) {
+					t.printStackTrace();
+					error[0] = t;
+				}
+			}
+		};
+		runner.start();
+
+		if (!latch.isTriggered()) {
+			latch.await();
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java]: Changing to
```
if (!latch.isTriggered()) {
			latch.await(2, TimeUnit.MINUTES);
		}

		assertTrue("Latch was not triggered within the given timeout.", latch.isTriggered());
```

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+		}
+
+		consumerOperator.cancel();
+		consumerOperator.close();
+
+		runner.interrupt();
+		runner.join();
+
+		assertNull(error[0]);
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java]: Changing to
```
		assertNull("Got error: " + error[0], error[0]);
```

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+
+		if (!latch.isTriggered()) {
+			latch.await();
+		}
+
+		consumerOperator.close();
+		runner.interrupt();
+		runner.join();
+
+		assertNull(error[0]);
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java]: See above

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java [hunk 12]
```diff
@@ -0,0 +1,491 @@
...
+
+		if (!latch.isTriggered()) {
+			latch.await();
+		}
+
+		consumerOperator.close();
+
+		runner.join();
+
+		assertNull(error[0]);
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java]: See above

File: flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java (hunk 8 above)
Comment (zentol) [inline flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java]: the line below has a `.` before fileName, (tmp variable) and I was trying to cover both cases with 1 comment by including `[.]` as some kind of optional part of the name.

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+				latch.trigger();
+				assertEquals(state1, map);
+				verifiedState[0] = true;
+				return null;
+			}
+		}).when(fetcher).restoreOffsets(anyMapOf(KafkaTopicPartition.class, Long.class));
+
+
+		final List<KafkaTopicPartition> partitions = new ArrayList<>();
+		partitions.add(new KafkaTopicPartition("abc", 13));
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java]: Replacing by
```
final static HashMap<KafkaTopicPartition, Long> PARTITION_STATE = new HashMap<>();

static {
    PARTITION_STATE.put(new KafkaTopicPartition("abc", 13), 16768L);
    PARTITION_STATE.put(new KafkaTopicPartition("def", 7), 987654321L);
}
```
 and
```
final List<KafkaTopicPartition> partitions = new ArrayList<>(PARTITION_STATE.keySet());
```
when creating the partitions list

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java [hunk 14]
```diff
@@ -0,0 +1,491 @@
...
+
+		// run the source asynchronously
+		Thread runner = new Thread() {
+			@Override
+			public void run() {
+				try {
+					consumerFunction.run(new DummySourceContext() {
+						@Override
+						public void collect(String element) {
+							//latch.trigger();
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseFrom12MigrationTest.java]: Fixing

File: flink-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFrom12MigrationTest.java (hunk 1 above)
Comment (aljoscha) [inline flink-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkFrom12MigrationTest.java]: Fixing

**Comment by aljoscha** [comment]
@zentol Thanks again for your thorough review!

What do you think about my earlier comment regarding unification:
```
About unifying the functionality of the tests/ITCases, I was planning to to that for the upcoming *From13 tests but for the *From11 and *From12 it's difficult because quite some stuff changed between 1.1 and 1.2 and I want the code to mostly work on all branches. For example, this is the new *From12 tests rebased on the master branch: https://github.com/aljoscha/flink/tree/jira-5969-backwards-compat-12-13. By transplanting but keeping the snapshots that were generated on Flink 1.2.0 we get backwards compatibility tests while the tests in this PR test compatibility within Flink 1.2.x.

```

File: flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java (hunk 6 above)
Comment (aljoscha) [inline flink-fs-tests/src/test/java/org/apache/flink/hdfstests/ContinuousFileProcessingFrom12MigrationTest.java]: It never get's unblocked and just goes away when the test finishes.

**Comment by zentol** [comment]
@aljoscha Having the same code on all branches is rather nice; never looked at it from that angle before. Not modifying existing tests is a plus anyway; so trying to unify from here on out sounds like a good idea.

**Comment by aljoscha** [comment]
Yep, though it is code duplication both across time (releases) and versions, now. 😅 I definitely have plans to unify for the `*From13MigrationTests`, though, since the interface is a bit more stable starting with 1.2.

**Comment by aljoscha** [comment]
Alright, so we're good to merge?

**Comment by zentol** [comment]
yes, +1 from my side.

**Comment by aljoscha** [comment]
Manually merged

**Comment by aljoscha** [comment]
@zentol Thanks for reviewing! 😃 

I merged this also on master, so that we have backwards compatibility tests from 1.2 to the current master, which was the reason for this whole exercise.

**Pull request closed without merging**